# Практическая работа №2: Исследование алгоритмов формирования аддитивных цепочек

Выполнила студентка гр. 1303 Андреева Елизавета. Вариант №1.

## Цель работы



Формирование представления о аддитивных цепочках, выработка умения составлять и применять алгоритмы для нахождения минимальных аддитивных цепочек для заданного числа, прививание навыка использования систем компьютерной алгебры для реализации алгоритмов.

## Основные теоретические положения


### Определение аддитивной цепочки

Аддитивной цепочкой для $n \in N$ называется последовательность $1=a_1, a_2, a_3, \dots, a_m=n$, где $a_i=a_j+a_k$, $k\leq j<i$ $\forall i=1..m$

$l(n)=m$ - минимальная аддитивная цепочка

Для $n \leq 1000$ справедлива следующая рекуретная формула:
$$
l(n) = min\{l(n - 1) + 1, l_n\} - \sigma_n \\
l_p= \begin{cases} \infty,\ если\ p -\ простое \\ l(q)+l(^p/_q),\ если\ q -\ минимальный\ простой\ делитель\ p \end{cases}\\
\sigma_n = {0,1}
$$
Для $\sigma_n$ нет формулы, для чисел меньших 200 $\sigma_n$ равняется единице при следующих значениях: $23, 43, 59, 77, 83, 107, 149, 163, 165, 179, 203$

### Типы шагов аддитивной цепочки
   Если $a_i < a_{i+1}$, то равенство $a_i = a_j+a_k$, $i = 1..m$, $0\leqslant k\leqslant j < i$ называется шагом аддитивной цепочки.
 
   Шаги:
   - Удвоение, если j = k = i - 1  $a_i = 2 * a_{i-1}$,
    
   - Звёздный, если j = i - 1, k = 0..i - 1,
    
   - Малый шаг, если $ \lambda (a_i)$ = $ \lambda (a_{i-1})$

### Алгоритм Брауэра
   Алгоритм Брауера вычисляет n-ную стпенень за $\lambda(n)+\frac{(1+o(1))\lambda(n)}{\lambda(\lambda(n))}$ операций.

   Для некоторых n, k Брауэрские цепочки задаются в виде рекурентной формулы: 
   $B_k(n) =\begin{cases}1, 2, 3, ... 2^k-1\text{, если } n<2^k \\
   B_k(q), 2q, 4q, 8q, ..., 2^kq, n,\text{ если } n \geqslant 2^k\ \text{и} q = \lfloor\frac{n}{2^k}\rfloor \end{cases}$,


### Звездная цепочка. Гипотеза Шольца-Брауэра

Звездная цепочка - это аддитивная цепочка, в которой каждый элемент получен при помощи сложения предыдущего и какого-то ещё: $a_i=a_{i-1}+a_k$, где $0 \leq k \leq i-1$

$l^*(n)$ - минимальная звездная цепочка. Очевидно, что $l^*(n) \geq l(n)$

Гипотеза Шольца-Брауэра заключается в следующем: $l^*(2^n-1) \geq l^*(n)+n-1$

### Алгоритм дробления индексов

Алгоритм дробления индексов находит минимальную звездную цепочку для числа $n$.

Введём опредление вектора индексов. Вектором индексов аддитивной цепочки называется последовательность $\{ r_i \}_{i=1}^{m}$, где $r_i = {x: 1 \leq x \leq i}$, такая то $a_i = a_{i-1} + a_{r_i}, \, i=2..m$

Наибольшая звездная цепочка: $S_{max}=\{1, 2, 4, \dots, 2^m\} <=> r_{max} = \{1, 2, 3, \dots, m\}$
Наименьшая звездная цепочка: $S_{min}=\{1,2,3,\dots , m\} <=> r_{min}=\{1, 1, 1, \dots, 1\}$

Будем считать, что $r=\{ r_i \}_{i=1}^{m}$ больше $\tilde{r}=\{ \tilde{r}_i \}_{i=1}^{m}$ если $r_1=\tilde{r}_1; r_2=\tilde{r}_2; \dots; r_{i-1}=\tilde{r}_{i-1};r_i>\tilde{r}_i$

Нахождение минимальной аддитивной цепочки является переборной задачей. Для начала будем каждый вектор индексов разбивать на два: $r=\{ r_i \}_{i=1}^{q}$ - фиксированный вектор, $\rho=\{ \rho_j \}_{j=q+1}^{m}$ - меняющийся вектор. Заметим, что для фиксированного вектора можно вычислить максимальное и минимальное значение аддитивной цепочки:
$$
a_{max}: \{ r_i \}_{i=1}^{q} \cup \{ q+1, q+2, \dots, m \} => a_{max} = a_{q+1} \cdot 2^{m-q}\\
a_{min}: \{ r_i \}_{i=1}^{q} \cup \{ 1, 1, \dots, 1 \} => a_{min}=a_{q+1}+m-q
$$

Алгоритм перебора будет заключаться в следующем: текущий вектор индексов разбивается по середине на фиксированный и изменющийся. Берётся фиксированная часть, вычисляются $a_{max}$ и $a_{min}$. Если $n \notin [a_{min}, a_{max}]$, то осуществляется переход к следующему фиксированному вектору(в сторону уменьшения или увеличения). Если $n$ попало в данный промежуток, то осуществляется перебор по меняющемуся вектору. Если для заданной длины не нашлось аддитивной цепочки, то увеличить длину аддитивной цепочки. Длина аддитивной цепочки $m \in [\underline{l}(n), \overline{l}(n)]$, где $\underline{l}(n)=\lceil log_2 n \rceil,\, \overline{l}(n)=\lambda(n)+\nu(n)-1$

## Постановка задачи

Реализовать точные и приближённые алгоритмы нахождения минимальных аддитивных цепочек с использованием системы компьютерной алгебры SageMath, провести анализ алгоритмов. Полученные результаты содержательно проинтерпретировать.


## Выполнение работы

### Алгоритм Брауэра

Создадим функцию реализующую алгоритм Брауэра.
Функция brauer реализует алгоритм Брауэра, принимает в качестве аргументов число n, для которого нужно найти аддитивную цепочку, и параметр k.

In [8]:
def brauer(n, k):
    pow_2 = pow(2, k)
    r_list, q_list, q = [], [], n
    
    while q >= pow_2:
        temp_q = q // pow_2
        q_list.append(temp_q)
        temp_r = q % pow_2
        r_list.append(temp_r)
        q = temp_q
        
    chain = []
    nums = []
    
    for i in range(1, pow_2):
        nums.append(i)
    
    chain_part = q_list[len(q_list) - 1]
    
    for i in range(len(q_list) - 1, -1, -1):
        for j in range(k):
            chain_part *= 2
            chain.append(chain_part)
        chain_part += r_list[i]
        if r_list[i] != 0:
            chain.append(chain_part)
            
    while chain[0] < pow_2:
        chain.pop(0)
            
    return chain

k_list = [2, 3, 4]
n_list = {    17:[1,2,4,8,16,17], 
              119:[1, 2, 3, 5, 7, 14, 28, 56, 112, 119],
              139:[1, 2, 4, 8, 9, 17, 34, 68, 69, 138, 139],
              234:[1,2,4,5,7,14,28,56,112,117,234], 
              511:[1,2,3,6,12,15,30,60,120,240,255,510,511], 
              1028:[1,2,4,8,16,32,64,128,256,257,514,1028], 
              1001:[1,2,4,5,10,15,30,60,120,125,250,500,1000,1001], 
              1002:[1,2,4,5,10,15,30,60,120,125,250,500,501,1002]}
for k in k_list:
    for n in n_list:
        temp_res = brauer(n, k)
        print(f"k = {k}, n = {n}")
        print(f"Цепочка Брауэра: {temp_res}, Длина = {len(temp_res)}\n Минимальная цепочка: {n_list[n]}, Длина = {len(n_list[n])}\n")


k = 2, n = 17
Цепочка Брауэра: [4, 8, 16, 17], Длина = 4
 Минимальная цепочка: [1, 2, 4, 8, 16, 17], Длина = 6

k = 2, n = 119
Цепочка Брауэра: [4, 7, 14, 28, 29, 58, 116, 119], Длина = 8
 Минимальная цепочка: [1, 2, 3, 5, 7, 14, 28, 56, 112, 119], Длина = 10

k = 2, n = 139
Цепочка Брауэра: [4, 8, 16, 32, 34, 68, 136, 139], Длина = 8
 Минимальная цепочка: [1, 2, 4, 8, 9, 17, 34, 68, 69, 138, 139], Длина = 11

k = 2, n = 234
Цепочка Брауэра: [6, 12, 14, 28, 56, 58, 116, 232, 234], Длина = 9
 Минимальная цепочка: [1, 2, 4, 5, 7, 14, 28, 56, 112, 117, 234], Длина = 11

k = 2, n = 511
Цепочка Брауэра: [4, 7, 14, 28, 31, 62, 124, 127, 254, 508, 511], Длина = 11
 Минимальная цепочка: [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511], Длина = 13

k = 2, n = 1028
Цепочка Брауэра: [4, 8, 16, 32, 64, 128, 256, 257, 514, 1028], Длина = 10
 Минимальная цепочка: [1, 2, 4, 8, 16, 32, 64, 128, 256, 257, 514, 1028], Длина = 12

k = 2, n = 1001
Цепочка Брауэра: [6, 12, 15, 30, 60, 62, 124, 248, 25

| n | k | Длина цепочки Брауэра | Аддитивная цепочка Брауэера | Длина минимальной цепочки | Минимальная цепочка |
|---|---|-----------------------|-----------------------------|---------------------------|---------------------|
| 17 | 2  | 7 | [1, 2, 3, 4, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] |
| 17 | 3  | 10 | [1, 2, 3, 4, 5, 6, 7, 8, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] |
| 17 | 4  | 17 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17] | 6 | [1, 2, 4, 8, 16, 17] |
| 119 | 2  | 11 | [1, 2, 3, 4, 7, 14, 28, 29, 58, 116, 119] | 10 | [1, 2, 3, 5, 7, 14, 28, 56, 112, 119] |
| 119 | 3  | 13 | [1, 2, 3, 4, 5, 6, 7, 8, 14, 28, 56, 112, 119] | 10 | [1, 2, 3, 5, 7, 14, 28, 56, 112, 119] |
| 119 | 4  | 19 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 28, 56, 112, 119] | 10 | [1, 2, 3, 5, 7, 14, 28, 56, 112, 119] |
| 139 | 2  | 11 | [1, 2, 3, 4, 8, 16, 32, 34, 68, 136, 139] | 11 | [1, 2, 4, 8, 9, 17, 34, 68, 69, 138, 139] |
| 139 | 3  | 14 | [1, 2, 3, 4, 5, 6, 7, 8, 16, 17, 34, 68, 136, 139] | 11 | [1, 2, 4, 8, 9, 17, 34, 68, 69, 138, 139] |
| 139 | 4  | 19 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 28, 56, 112, 119] | 11 | [1, 2, 4, 8, 9, 17, 34, 68, 69, 138, 139] |
| 234 | 2  | 12 | [1, 2, 3, 6, 12, 14, 28, 56, 58, 116, 232, 234] | 11 | [1, 2, 4, 5, 7, 14, 28, 56, 112, 117, 234] |
| 234 | 3  | 15 | [1, 2, 3, 4, 5, 6, 7, 8, 12, 24, 29, 58, 116, 232, 234] | 11 | [1, 2, 4, 5, 7, 14, 28, 56, 112, 117, 234] |
| 234 | 4  | 20 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 28, 56, 112, 224, 234] | 11 | [1, 2, 4, 5, 7, 14, 28, 56, 112, 117, 234] |
| 511 | 2  | 14 | [1, 2, 3, 4, 7, 14, 28, 31, 62, 124, 127, 254, 508, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 511 | 3  | 15 | [1, 2, 3, 4, 5, 6, 7, 14, 28, 56, 63, 126, 252, 504, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |
| 511 | 4  | 22 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 31, 62, 124, 248, 496, 511] | 13 | [1, 2, 3, 6, 12, 15, 30, 60, 120, 240, 255, 510, 511] |\n",
| 1028 | 2  | 12 | [1,2, 3, 4, 8, 16, 32, 64, 128, 256, 257, 514, 1028] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 257, 514, 1028] |
| 1028 | 3  | 16 | [1, 2, 3, 4, 5, 6, 7, 8, 16, 32, 64, 128, 256, 512, 1024, 1028] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 257, 514, 1028] |
| 1028 | 4  | 23 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 32, 64, 128, 256, 512, 1024, 1028] | 12 | [1, 2, 4, 8, 16, 32, 64, 128, 256, 257, 514, 1028] |
| 1001 | 2  | 15 | [1, 2, 3, 6, 12, 15, 30, 60, 62, 124, 248, 250, 500, 1000, 1001] | 14 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] |
| 1001 | 3  | 17 | [1, 2, 3, 4, 5, 6, 7, 8, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] | 14 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] |
| 1001 | 4  | 23 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 62, 124, 248, 496, 992, 1001] | 14 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 1000, 1001] |
| 1002 | 2  | 15 | [1, 2, 3, 6, 12, 15, 30, 60, 62, 124, 248, 250, 500, 1000, 1002] | 14 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 501, 1002] |
| 1002 | 3  | 17 | [1, 2, 3, 4, 5, 6, 7, 8, 15, 30, 60, 120, 125, 250, 500, 1000, 1002] | 14 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 501, 1002] |
| 1002 | 4  | 23 | [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 24, 48, 62, 124, 248, 496, 992, 1002] | 14 | [1, 2, 4, 5, 10, 15, 30, 60, 120, 125, 250, 500, 501, 1002] |

Из результатов заметим, что длина аддитивной цепочки, полученная методом Брауэра, отличается от длины минимальной цепочки. Также алгоритм Брауэра работает довольно быстро.

### Алгоритм дробления вектора индексов

Реализован алгоритм дробления индексов для нахождения минимальной аддитивной цепочки для числа $n$:

In [9]:
import time

def get_chain(r_i, ro_i):
    q = len(r_i)
    chain = [1]
    for i in r_i:
        chain.append(chain[-1] + chain[i - 1])
    for i in ro_i:
        chain.append(chain[-1] + chain[i - 1])
    return chain
    
def inc_index_vector(ind_vec, start_index=1):
    cur_max_index = start_index + len(ind_vec) - 1
    for i in range(1, len(ind_vec) + 1):
        ind_vec[-i] += 1
        if ind_vec[-i] > cur_max_index:
            if i == len(ind_vec):
                ind_vec.append(1)
                break
            ind_vec[-i] = 1
            cur_max_index -= 1
        else:
            break
    
def split_index_vector(n):
    count = 0
    m = int(log(n, 2))
    up_border = m + bin(n).count('1') - 1
    if m == log(n, 2):
        m -= 1
    while m <= up_border:
        q = int(m/2)
        if q == 0:
            q = 1
        r_i = [1 for i in range(q)]
        while len(r_i) == q:
            ro_i = [1 for i in range(m - q)]
            chain = get_chain(r_i, ro_i)
            if chain[-1] == n:
                return chain
            a_max = chain[q] * 2 ** (m - q)
            a_min = chain[q] + m - q
            if n < a_min or n > a_max:
                inc_index_vector(r_i)
                continue
            while len(ro_i) == m - q:
                chain = get_chain(r_i, ro_i)
                if chain[-1] == n:
                    return chain
                inc_index_vector(ro_i, q + 1)
            inc_index_vector(r_i)
        m += 1
            

def calc_time(time):
        print(f"Время выполнения: {time} секунд")

n_list = {    17:[1,2,4,8,16,17], 
              119:[1, 2, 3, 5, 7, 14, 28, 56, 112, 119],
              139:[1, 2, 4, 8, 9, 17, 34, 68, 69, 138, 139],
              234:[1,2,4,5,7,14,28,56,112,117,234], 
              511:[1,2,3,6,12,15,30,60,120,240,255,510,511]}
for n in n_list:
    print(f"n = {n}")
    time1 = time.time()
    add_chain = split_index_vector(n)
    time2 = time.time()
    calc_time(time2 - time1)
    print("Аддитивная цепочка:", add_chain)
    print("Длина аддитивной цепочки:", len(add_chain))

n = 17
Время выполнения: 0.006695270538330078 секунд
Аддитивная цепочка: [1, 2, 4, 8, 9, 17]
Длина аддитивной цепочки: 6
n = 119
Время выполнения: 0.46953654289245605 секунд
Аддитивная цепочка: [1, 2, 3, 4, 7, 14, 28, 56, 63, 119]
Длина аддитивной цепочки: 10
n = 139
Время выполнения: 5.3536272048950195 секунд
Аддитивная цепочка: [1, 2, 3, 4, 7, 10, 17, 34, 68, 71, 139]
Длина аддитивной цепочки: 11
n = 234
Время выполнения: 7.009759187698364 секунд
Аддитивная цепочка: [1, 2, 3, 5, 7, 14, 28, 56, 61, 117, 234]
Длина аддитивной цепочки: 11
n = 511
Время выполнения: 829.3554537296295 секунд
Аддитивная цепочка: [1, 2, 3, 4, 7, 14, 21, 42, 63, 126, 252, 259, 511]
Длина аддитивной цепочки: 13


В таблице приведены минимальные аддитивные цепочки для числа $n$, их длина и время нахождения алгоритмом дробления вектора индексов.

| $n$ | Длина | Время нахождения | Аддитивная цепочка |
|-----|-------|------------------|--------------------|
| 17  | 6 | 0.006695270538330078 секунд | [1, 2, 4, 8, 9, 17] |
| 119 | 10 | 0.46953654289245605 секунд | [1, 2, 3, 4, 7, 14, 28, 56, 63, 119] |
| 139 | 11 | 5.3536272048950195 секунд | [1, 2, 3, 4, 7, 14, 28, 56, 63, 119] |
| 234 | 11 | 7.009759187698364 секунд| [1, 2, 3, 5, 7, 14, 28, 56, 61, 117, 234] |
| 511 | 13 | 829.3554537296295 секунд | [1, 2, 3, 4, 7, 14, 21, 42, 63, 126, 252, 259, 511] |

В таблице представлено сравнение длин аддитивных цепочек для $n$ различными алгоритмами

| $n$  | Алгоритм дробления индексов | Алгоритм Брауэра |
| :--: | :-------------------------: | :----------: |
| 17   |           6                 |     7        |
| 119  |           10                |     11       |
| 139  |           11                |     11       |
| 234  |           11                |     12       |
| 511  |           13                |     14       |

---

#### Вывод

Как видно из первой таблицы, для нахождения наименьшей аддитивной цепочки в большинстве случаев требуется несколько десятков минут, если $n > 500$.
Можно заключить, что алгоритм дробления вектора индексов хорош для вычислений заранее, однако если требуется находить быстро, но не минимальную аддитивную цепочку, то лучше выбрать другой алгоритм, например, алгоритм Брауэра.

### Гипотеза Шольца-Брауэра

Проверим гипотезу Шольца-Брауэра, которая утверждает, что $l^*(2^n-1) \geq l^*(n)+n-1$.
Ниже представлена программа, которая реализует проверку справедливости гипотезы Шольца-Брауэра.

In [ ]:
def get_chain(r_i, ro_i):
    q = len(r_i)
    chain = [1]
    for i in r_i:
        chain.append(chain[-1] + chain[i - 1])
    for i in ro_i:
        chain.append(chain[-1] + chain[i - 1])
    return chain
    
def inc_index_vector(ind_vec, start_index=1):
    cur_max_index = start_index + len(ind_vec) - 1
    for i in range(1, len(ind_vec) + 1):
        ind_vec[-i] += 1
        if ind_vec[-i] > cur_max_index:
            if i == len(ind_vec):
                ind_vec.append(1)
                break
            ind_vec[-i] = 1
            cur_max_index -= 1
        else:
            break
    
def split_index_vector(n):
    count = 0
    m = int(log(n, 2))
    up_border = m + bin(n).count('1') - 1
    if m == log(n, 2):
        m -= 1
    while m <= up_border:
        q = int(m/2)
        if q == 0:
            q = 1
        r_i = [1 for i in range(q)]
        while len(r_i) == q:
            ro_i = [1 for i in range(m - q)]
            chain = get_chain(r_i, ro_i)
            if chain[-1] == n:
                return chain
            a_max = chain[q] * 2 ** (m - q)
            a_min = chain[q] + m - q
            if n < a_min or n > a_max:
                inc_index_vector(r_i)
                continue
            while len(ro_i) == m - q:
                chain = get_chain(r_i, ro_i)
                if chain[-1] == n:
                    return chain
                inc_index_vector(ro_i, q + 1)
            inc_index_vector(r_i)
        m += 1

        
for n in range(2, 13):
    left = len(split_index_vector(2 ** n -  1))
    right = len(split_index_vector(n)) + n - 1
    print(f"n={n}: {left} <= {right}")

n=2: 3 <= 3
n=3: 5 <= 5
n=4: 6 <= 6
n=5: 8 <= 8
n=6: 9 <= 9
n=7: 11 <= 11


В таблице приведены значения $l^*(2^n-1)$ и $l^*(n) + n - 1$ для $n \in (1, 10]$

| Значение $n$ | $$l^*(2^n-1)$$ | $$l^*(n) + n - 1$$ |
| :----------: | :----------: | :--------------: |
| 2 | 3 | 3 |
| 3 | 5 | 5 |
| 4 | 6 | 6 |
| 5 | 8 | 8 |
| 6 | 9 | 9 |
| 7 | 11 | 11 |
| 8 | 11 | 11 |
| 9 | 13 | 13 |
| 10 | 14 | 14 |
| 11 | 16 | 16 |
| 12 | 16 | 16 |

---

#### Вывод

Как видно из полученных данных, гипотеза Шольца-Брауэра справедлива для $n \leq 12$. Нахождение минимальной аддитивной цепочки для числа $n=2^k-1$, где $k \in N$ является очень затратной операцией. Благодаря этой гипотезе можно относительно быстро вычислить минимальную длину, по которой начать делать перебор для нахождения минимальной аддитивной цепочки.

## Выводы

---

Были изучены, реализованы и исследованы алгоритмы формирования аддитивных цепочек. В практической работе были рассмотрены следующие алгоритмы:
* Алгоритм Брауэра
* Алгоритм дробления вектора индексов

Алгоритм Брауэра не всегда вычисляют минимальную аддитивную цепочку, однако делает достаточно быстро. Алгоритм дробления вектора индексов является переборным алгоритмом для поиска наименьшей аддитивной цепочки для числа $n$. Так как это переборная задача, то время выполнения растёт быстрее, чем экспоненциальное. Поэтому если необходимо быстро, но не обязательно минимально вычислять аддитивную цепочку, то стоит выбрать приближённый алгоритм.

Также была проверена гипотеза Шольца-Брауэра на $n \in (1, 10]$. Благодаря этой гипотезе можно достаточно быстро узнать минимальную длину аддитивной цепочки для $n=2^k-1,\,k \in N$, и уже на этой длине осуществлять переборный алгоритм.